File: src/wbjp/strategy/samples/sma_cross.py
"""移動平均クロス戦略（順張り）。

短期移動平均が長期移動平均を上抜けたら買い、下抜けたら売り。
最も古典的なトレンドフォロー。

性格: トレンドが続く相場で強く、横ばいでは往復ビンタを食らう。
逆張り戦略と組み合わせると相互に弱点を補える。
"""

from __future__ import annotations

from typing import ClassVar

import polars as pl

from wbcore.domain.models import Signal
from wbcore.indicators.ohlcv import sma
from wbjp.strategy.base import IndicatorStrategy, StrategyContext


class SmaCrossStrategy(IndicatorStrategy):
    name: ClassVar[str] = "sma_cross"

    def __init__(self, fast: int = 25, slow: int = 75) -> None:
        if fast >= slow:
            raise ValueError(f"fast は slow より小さく: fast={fast}, slow={slow}")
        self.fast = fast
        self.slow = slow
        # 長期線が確定した翌日から前日比較ができるので +1 本
        self.warmup_bars = slow + 1

    def indicators(self) -> list[pl.Expr]:
        return [sma(self.fast), sma(self.slow)]

    def evaluate(self, symbol: str, frame: pl.DataFrame, ctx: StrategyContext) -> Signal | None:
        fast_col, slow_col = f"sma_{self.fast}", f"sma_{self.slow}"
        recent = frame.tail(2)
        if recent.height < 2:
            return None

        prev, current = recent.row(0, named=True), recent.row(1, named=True)
        if any(row[col] is None for row in (prev, current) for col in (fast_col, slow_col)):
            return None

        prev_diff = prev[fast_col] - prev[slow_col]
        current_diff = current[fast_col] - current[slow_col]

        # 乖離率を確信度に使う。離れているほどトレンドが明確とみなす。
        confidence = min(1.0, abs(current_diff) / current[slow_col] * 20.0)

        if prev_diff <= 0 < current_diff:
            return Signal(
                self.name,
                symbol,
                direction=1.0,
                confidence=max(confidence, 0.3),
                reason=f"{self.fast}日線が{self.slow}日線を上抜け",
                meta={"fast": current[fast_col], "slow": current[slow_col]},
            )
        if prev_diff >= 0 > current_diff:
            return Signal(
                self.name,
                symbol,
                direction=-1.0,
                confidence=max(confidence, 0.3),
                reason=f"{self.fast}日線が{self.slow}日線を下抜け",
                meta={"fast": current[fast_col], "slow": current[slow_col]},
            )

        # クロスしていない間も、どちら側にいるかは意見として出す。
        # これが無いと、クロスした当日しか保有を維持できない。
        direction = 1.0 if current_diff > 0 else -1.0
        return Signal(
            self.name,
            symbol,
            direction=direction,
            confidence=confidence * 0.5,
            reason=f"{self.fast}日線が{self.slow}日線の{'上' if direction > 0 else '下'}で推移",
            meta={"fast": current[fast_col], "slow": current[slow_col]},
        )

    def describe(self) -> str:
        return f"{self.name}(fast={self.fast}, slow={self.slow})"

File: pkg/wbjp/strategy/sma_cross.go
package strategy

import (
	"fmt"
	"math"

	"github.com/lovemoneyhotspring/jstock-go/pkg/wbcore/domain"
)

// SMACross は移動平均クロス戦略（順張り）。
//
// 短期移動平均が長期移動平均を上抜けたら買い、下抜けたら売り。
// トレンドが続く相場で強く、横ばいでは往復ビンタを食らう。逆張り戦略と
// 組み合わせると相互に弱点を補える。
type SMACross struct {
	Fast int
	Slow int

	warmup int
}

// NewSMACross は移動平均クロス戦略を作る。
func NewSMACross(fast, slow int) (*SMACross, error) {
	if fast <= 0 {
		fast = 25
	}
	if slow <= 0 {
		slow = 75
	}
	if fast >= slow {
		return nil, fmt.Errorf("fast は slow より小さく: fast=%d, slow=%d", fast, slow)
	}
	// 長期線が確定した翌日から前日比較ができるので +1 本
	return &SMACross{Fast: fast, Slow: slow, warmup: slow + 1}, nil
}

func (s *SMACross) Name() string    { return "sma_cross" }
func (s *SMACross) WarmupBars() int { return s.warmup }
func (s *SMACross) Describe() string {
	return fmt.Sprintf("%s(fast=%d, slow=%d)", s.Name(), s.Fast, s.Slow)
}

func (s *SMACross) OnBars(ctx *Context) ([]domain.Signal, error) {
	return eachSymbol(ctx, s.warmup, s.evaluate), nil
}

func (s *SMACross) evaluate(symbol string, v View) *domain.Signal {
	fast := v.SMA(s.Fast)
	slow := v.SMA(s.Slow)

	currFast, currSlow := last(fast), last(slow)
	prevFast, prevSlow := ago(fast, 1), ago(slow, 1)
	if anyNaN(currFast, currSlow, prevFast, prevSlow) || currSlow == 0 {
		return nil
	}

	prevDiff := prevFast - prevSlow
	currDiff := currFast - currSlow

	// 乖離率を確信度に使う。離れているほどトレンドが明確とみなす。
	confidence := math.Min(1.0, math.Abs(currDiff)/currSlow*20.0)
	meta := map[string]any{"fast": currFast, "slow": currSlow}

	if prevDiff <= 0 && currDiff > 0 {
		return signal(s.Name(), symbol, 1.0, math.Max(confidence, 0.3),
			fmt.Sprintf("%d日線が%d日線を上抜け", s.Fast, s.Slow), meta)
	}
	if prevDiff >= 0 && currDiff < 0 {
		return signal(s.Name(), symbol, -1.0, math.Max(confidence, 0.3),
			fmt.Sprintf("%d日線が%d日線を下抜け", s.Fast, s.Slow), meta)
	}

	// クロスしていない間も、どちら側にいるかは意見として出す。
	// これが無いと、クロスした当日しか保有を維持できない。
	direction, side := -1.0, "下"
	if currDiff > 0 {
		direction, side = 1.0, "上"
	}
	return signal(s.Name(), symbol, direction, confidence*0.5,
		fmt.Sprintf("%d日線が%d日線の%sで推移", s.Fast, s.Slow, side), meta)
}
